### PR TITLE
add authentication for user login and creation

### PR DESCRIPTION
**Pull request by dlimx**
Adds `passport` and `jsonwebtoken` based authentication

Adds `authService.jwt` as middleware for our future API routes

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,41 +3,48 @@
...
-    "react-scripts": "3.4.1"
+    "react-scripts": "3.4.1",
+    "yup": "^0.29.1"
   },
   "devDependencies": {
     "@testing-library/jest-dom": "^4.2.4",
     "@testing-library/react": "^9.3.2",
     "@testing-library/user-event": "^7.1.2",
     "acorn": "^7.3.1",
-    "babel-eslint": "^10.1.0",
```
Comment (DomPSU) [inline package.json:40]: I thought we were going to add babel-eslint?

File: src/apis/axios.js [hunk 2]
```diff
@@ -1,13 +1,10 @@
...
 
-// baseURL: replace with backend url 
+// baseURL: replace with backend url
 export default axios.create({
-	baseURL: 'http://localhost:3001',
-	headers: {
-		headerType: 'example header type'
-	}
-	
+  baseURL: isDevelopment() ? 'http://localhost:3001' : '/',
```
[the remark this answers is not in the document]
Comment (dlimx) [reply src/apis/axios.js:6]: Both should probably include the `/api` path - I'll add it in a followup PR that includes the front-end auth code

File: models/seeds.sql [hunk 3]
```diff
@@ -1,42 +1,103 @@
...
+        NOW(), NOW());
+INSERT INTO Users (userLevelID, email, password, firstName, lastName, createdAt, updatedAt)
+VALUES ((SELECT userLevelID FROM UserLevels WHERE role = 'buyer'), 'jrogan@gmail.com', 'password4', 'joe', 'rogan',
+        NOW(), NOW());
+INSERT INTO Users (userLevelID, email, password, firstName, lastName, createdAt, updatedAt)
+VALUES ((SELECT userLevelID FROM UserLevels WHERE role = 'cashier'), 'gfring@gmail.com', 'password5', 'gus', 'fring',
+        NOW(), NOW());
+INSERT INTO Users (userLevelID, email, password, firstName, lastName, createdAt, updatedAt)
+VALUES ((SELECT userLevelID FROM UserLevels WHERE role = 'cashier'), 'sgoodman@gmail.com', 'password6', 'saul',
+        'goodman', NOW(), NOW());
```
[the remark this answers is not in the document]
Comment (dlimx) [reply models/seeds.sql:36]: Good point

File: package.json (hunk 1 above)
Comment (dlimx) [reply package.json:40]: This was before our discussion

File: models/schema.sql [hunk 4]
```diff
@@ -49,7 +49,4 @@ CREATE TABLE Transactions(
     CONSTRAINT FOREIGN KEY (productID) REFERENCES Products(productID)
 );
 
-SET @@GLOBAL.auto_increment_increment = 10;
-SET @@SESSION.auto_increment_increment = 10;
-
```
[the remark this answers is not in the document]
Comment (dlimx) [reply models/schema.sql:54]: The idea was to not need to worry about increment amounts - usually we'd want to remove dead code as if we ever need it it's still in our history

File: models/user.js [hunk 5]
```diff
@@ -1,17 +1,63 @@
...
+const _ = require('lodash');
 const db = require('./db');
+const constants = require('../constants/constants');
 
+// remove password for getAll/getUserById to prevent password from being accessed
 const getAll = async () => {
-  return db.pool.asyncQuery('SELECT * FROM Users');
+  const data = await db.pool.asyncQuery('SELECT * FROM Users');
+  data.forEach((user) => {
+    // eslint-disable-next-line no-param-reassign
```
[the remark this answers is not in the document]
Comment (dlimx) [reply models/user.js:10]: We want to make `eslint` work for us - so we can disable rules one by one only in exceptional cases

File: models/user.js [hunk 6]
```diff
@@ -1,17 +1,63 @@
...
+      data.password,
+      data.firstName,
+      data.lastName,
+      now,
+      now,
+    ],
+  );
+  return db.pool.asyncQuery('SELECT * FROM Users WHERE email = ?', [
+    data.email,
+  ]);
```
[the remark this answers is not in the document]
Comment (dlimx) [reply models/user.js:54]: Sure

File: package.json (hunk 1 above)
Comment (DomPSU) [reply package.json:40]: okay was not sure.

File: models/user.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (DomPSU) [reply models/user.js:10]: Sounds good.

File: src/apis/axios.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (DomPSU) [reply src/apis/axios.js:6]: okay cool, thanks for clarifying.

File: models/user.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (DomPSU) [reply models/user.js:54]: Are you adding it to this pr or should I add it on my next one?

[the remark this answers is not in the document]
Comment (dlimx) [reply models/user.js:54]: I'll add it now!

File: package.json (hunk 1 above)
Comment (dlimx) [reply package.json:40]: Nah, it's a good callout, will re-add

File: models/schema.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (DomPSU) [reply models/schema.sql:54]: Actually I don't think I need it since you updated the inserts to use queries instead of the hard coded fks.

**Pull request merged**